lua/nvim-treesitter/refactor/smart_rename.lua

```diff
@@ -0,0 +1,75 @@
+-- Binds a keybinding to smart rename definitions and usages.
+-- Can be used directly using the `smart_rename` function.
+
+local ts_utils = require'nvim-treesitter.ts_utils'
+local configs = require'nvim-treesitter.configs'
+local api = vim.api
+
+local M = {}
+
+function M.smart_rename(bufnr)
+  local bufnr = bufnr or api.nvim_get_current_buf()
+  local node_at_point = ts_utils.get_node_at_cursor()
+
+  if not node_at_point then
+    print('No node to rename!')
+    return
+  end
+
+  local node_text = ts_utils.get_node_text(node_at_point)[1]
+  local new_name = vim.fn.input('New name: ', node_text or '')
+
+  -- Empty name cancels the interaction or ESC
+  if not new_name or #new_name < 1 then return end
+
+  local definition, scope = ts_utils.find_definition(node_at_point, bufnr)
+  local nodes_to_rename = ts_utils.find_usages(node_at_point, scope)
+
+  if not vim.tbl_contains(nodes_to_rename, node_at_point) then
+    table.insert(nodes_to_rename, node_at_point)
+  end
+
+  if definition and not vim.tbl_contains(nodes_to_rename, definition) then
+    table.insert(nodes_to_rename, definition)
+  end
+
+  if #nodes_to_rename < 1 then
+    print('No nodes to rename!')
```

Comment:
same here, should we output this as a warning ?

Reply:
Not sure if that can happen though, as we for sure rename something, at least the node at cursor

Reply:
I'll remove it if it's not a possible scenario

Reply:
Let's keep it to be safe